### PR TITLE
disable optimizers setup during testing

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Disable configuring optimizers and schedulers during testing.

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by SkafteNicki** [comment]
@rohitgr7 PR is looking good, just wanted to ask why it is necessary to disable configuration during testing?

**Comment by rohitgr7** [comment]
@SkafteNicki if someone wants to use lightning module for the sole purpose of testing or monitoring the performance of their model, then only test part of lightning module is required so I don't think optimizers or schedulers are required here. If we keep it as it is, I might throw an error since .test actually calls .fit

**Comment by SkafteNicki** [comment]
@rohitgr7 good point. If there is a use case right now on master that produce an error (related to this), we should probably write an test that quantifies that this PR solves that :]

**Comment by rohitgr7** [comment]
@SkafteNicki a simple one I can think of is just implement test_step and test_epoch_end and call trainer.test(). Ideally it should work but it throws an error. We definitely need some tests for this. Marked as WIP.

**Comment by rohitgr7** [comment]
lot's of refactoring going on with accelerators. I guess I'll wait for some time to work on this PR.

File: pytorch_lightning/plugins/apex.py [hunk 1]
```diff
@@ -23,10 +23,12 @@ class ApexPlugin:
     def __init__(self, trainer):
         self.trainer = trainer
 
-    def connect(self, model, optimizers):
+    def connect(self, model, optimizers=None):
         model, optimizers = self.configure_apex(model, optimizers, self.trainer.amp_level)
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/apex.py:27]: is `LightningModule.configure_apex` not used anymore??

**Comment by rohitgr7** [comment]
should I write supporting tests for each backend?

File: pytorch_lightning/accelerators/ddp_backend.py [hunk 2]
```diff
@@ -139,11 +139,6 @@ def _check_can_spawn_children(self):
                 " This is not supported in DDP mode, switch to `distributed_backend='ddp_spawn'` instead."
             )
 
-    def set_world_ranks(self, process_idx):
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/ddp_backend.py:142]: why is this removed?

Comment (rohitgr7) [reply pytorch_lightning/accelerators/ddp_backend.py]: my bad :sweat_smile:

Comment (rohitgr7) [reply pytorch_lightning/accelerators/ddp_backend.py]: thanks for finding it :)

File: pytorch_lightning/accelerators/cpu_backend.py [hunk 3]
```diff
@@ -32,10 +32,12 @@ def setup(self, model):
...
         # allow for lr schedulers as well
-        optimizers, lr_schedulers, optimizer_frequencies = self.trainer.init_optimizers(model)
-        self.trainer.optimizers = optimizers
-        self.trainer.lr_schedulers = lr_schedulers
-        self.trainer.optimizer_frequencies = optimizer_frequencies
+        if not self.trainer.testing:
+            optimizers, lr_schedulers, optimizer_frequencies = self.trainer.init_optimizers(model)
+            self.trainer.optimizers = optimizers
+            self.trainer.lr_schedulers = lr_schedulers
+            self.trainer.optimizer_frequencies = optimizer_frequencies
```
Comment (rohitgr7) [inline pytorch_lightning/accelerators/cpu_backend.py:39]: this is a repetitive code in every accelerator. Should be moved to the `Accelerator` class?

File: pytorch_lightning/plugins/native_amp.py [hunk 4]
```diff
@@ -19,8 +19,7 @@ class NativeAMP:
     def __init__(self, trainer):
         self.trainer = trainer
 
-    def connect(self, model, optimizers):
-        self.trainer.optimizers = optimizers
+    def connect(self, model, optimizers=None):
```
Comment (ananthsub) [inline pytorch_lightning/plugins/native_amp.py:22]: are there type annotations that can be added?

File: pytorch_lightning/accelerators/dp_backend.py [hunk 5]
```diff
@@ -65,17 +67,17 @@ def __init_torch_data_parallel(self, model):
         model = LightningDataParallel(model, device_ids=device_ids)
         return model
 
-    def __init_half_precision(self, model):
+    def _setup_half_precision(self, model):
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/dp_backend.py:70]: i'm curious, why the naming change? the pattern here seems to be `__init` everywhere?

**Review by ananthsub — APPROVED** [review]
I only have a few style comments, but the core looks good to me! Thank you for adding this @rohitgr7 !

File: pytorch_lightning/accelerators/dp_backend.py (hunk 5 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/dp_backend.py:70]: I changed it before rebasing the refactors to match with other backend APIs, but while rebasing it never showed up in conflict. Reverted it. Thanks again.

File: pytorch_lightning/plugins/native_amp.py (hunk 4 above)
Comment (rohitgr7) [reply pytorch_lightning/plugins/native_amp.py:22]: not sure if it's required for internal apis.

Comment (Borda) [reply pytorch_lightning/plugins/native_amp.py:22]: maybe not required but we shall add it everywhere at lias some basic (not all details)

File: pytorch_lightning/plugins/apex.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/plugins/apex.py:27]: probably moved due to refactoring, but it still stays

File: pytorch_lightning/accelerators/cpu_backend.py (hunk 3 above)
Comment (Borda) [reply pytorch_lightning/accelerators/cpu_backend.py:39]: this will be cleaned later, right? @williamFalcon

File: pytorch_lightning/plugins/apex.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/plugins/apex.py:27]: @williamFalcon

File: pytorch_lightning/accelerators/horovod_backend.py [hunk 6]
```diff
@@ -48,40 +48,42 @@ def setup(self, model):
...
-        def filter_named_parameters(model, optimizer):
-            opt_params = set([p for group in optimizer.param_groups for p in group.get('params', [])])
-            return [(name, p) for name, p in model.named_parameters() if p in opt_params]
-
-        # Horovod: wrap optimizers to perform gradient aggregation via allreduce
-        self.trainer.optimizers = [
-            hvd.DistributedOptimizer(optimizer, named_parameters=filter_named_parameters(model, optimizer))
-            for optimizer in self.trainer.optimizers
-        ]
+        if hasattr(self.trainer, 'optimizers'):
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/horovod_backend.py:59]: can we not initialize trainer.optimizers = [] from the start?

Comment (rohitgr7) [reply pytorch_lightning/accelerators/horovod_backend.py:59]: Yeah I thought about this, but then I thought it would be better to not have the optimizers and schedulers attributes at all during testing. I can change it to what you suggested, let me know.

Comment (awaelchli) [reply pytorch_lightning/accelerators/horovod_backend.py:59]: I think it would be cleaner. 
If your concern is that the user may access the trainer.optimizers by accident, well, they will see it is empty at test time so they get an error anyway.

File: pytorch_lightning/accelerators/cpu_backend.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/cpu_backend.py]: for now, I moved it to the Accelerator.

File: pytorch_lightning/trainer/__init__.py [hunk 7]
```diff
@@ -1095,4 +1095,3 @@ def tbptt_split_batch(self, batch, split_size):
 from pytorch_lightning.utilities.seed import seed_everything
 
 __all__ = ["Trainer", "seed_everything"]
-
```
Comment (awaelchli) [inline pytorch_lightning/trainer/__init__.py:1098]: can you remove all these whitespace changes so we get fewer files changed?

File: pytorch_lightning/accelerators/base_backend.py [hunk 8]
```diff
@@ -165,3 +165,9 @@ def on_train_epoch_end(self):
 
     def early_stopping_should_stop(self, pl_module):
         return self.trainer.should_stop
+
+    def setup_optimizers(self, model):
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/base_backend.py:169]: +1 for putting this in the base class. I like it. 
Have you thought about moving the "if self.testing" logic here as well?
It looks like it is repeated everywhere, so it could simply be handled here. what do you think?

**Review by awaelchli — APPROVED** [review]
overall looks good. I like it.
some unrelated changes made it in here, probably from rebasing/merging. 
make sure to minimize the number of whitespace changes in unrelated files.

File: pytorch_lightning/accelerators/base_backend.py (hunk 8 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/base_backend.py:169]: yep, good point.

File: pytorch_lightning/trainer/__init__.py (hunk 7 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/__init__.py:1098]: sure, just wanted to see all greens in tests locally. I'll remove them.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/accelerators/base_backend.py [hunk 9]
```diff
@@ -165,3 +166,10 @@ def on_train_epoch_end(self):
...
+        if not self.trainer.testing:
+            optimizers, lr_schedulers, optimizer_frequencies = self.trainer.init_optimizers(model)
+            self.trainer.optimizers = optimizers
+            self.trainer.lr_schedulers = lr_schedulers
+            self.trainer.optimizer_frequencies = optimizer_frequencies
```
Comment (Borda) [inline pytorch_lightning/accelerators/base_backend.py:175]:
```suggestion
        if self.trainer.testing:
            return
        optimizers, lr_schedulers, optimizer_frequencies = self.trainer.init_optimizers(model)
        self.trainer.optimizers = optimizers
        self.trainer.lr_schedulers = lr_schedulers
        self.trainer.optimizer_frequencies = optimizer_frequencies
```

File: pytorch_lightning/accelerators/ddp_base_backend.py [hunk 10]
```diff
@@ -13,20 +13,21 @@
...
-from pytorch_lightning.utilities import AMPType
-from pytorch_lightning.accelerators.base_backend import Accelerator
+import torch
 import torch.distributed as torch_distrib
 import torch.distributed as dist
-from pytorch_lightning.utilities.cloud_io import atomic_save
-from pytorch_lightning.utilities.distributed import rank_zero_warn, rank_zero_only
+
 from pytorch_lightning import _logger as log
+from pytorch_lightning.accelerators.base_backend import Accelerator
```
Comment (Borda) [inline pytorch_lightning/accelerators/ddp_base_backend.py:22]: this is good but rather make it in separate PR

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
seems some more GPU tests are failing, @rohitgr7 mind check it...

File: pytorch_lightning/accelerators/ddp_base_backend.py (hunk 10 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/ddp_base_backend.py:22]: this is just isort

File: pytorch_lightning/plugins/apex.py [hunk 11]
```diff
@@ -24,13 +24,8 @@ def __init__(self, trainer):
...
 
     def connect(self, model, optimizers):
-        model, optimizers = self.configure_apex(model, optimizers, self.trainer.amp_level)
-        self.trainer.optimizers = optimizers
-        self.trainer.reinit_scheduler_properties(self.trainer.optimizers, self.trainer.lr_schedulers)
-        return model, optimizers
-
-    def configure_apex(self, model, optimizers, amp_level):
-        model, optimizers = amp.initialize(model, optimizers, opt_level=amp_level)
+        model, optimizers = model.configure_apex(model, optimizers, self.trainer.amp_level)
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/apex.py:27]:
```suggestion
        model, optimizers = model.configure_apex(amp, model, optimizers, self.trainer.amp_level)
```

File: tests/base/model_valid_epoch_ends.py [hunk 12]
```diff
@@ -30,6 +30,7 @@ def _mean(res, key):
         results = {'progress_bar': metrics_dict, 'log': metrics_dict}
         return results
 
+
```
Comment (rohitgr7) [inline tests/base/model_valid_epoch_ends.py:33]:
```suggestion
```

File: pytorch_lightning/trainer/__init__.py [hunk 13]
```diff
@@ -1095,3 +1095,4 @@ def tbptt_split_batch(self, batch, split_size):
 from pytorch_lightning.utilities.seed import seed_everything
 
 __all__ = ["Trainer", "seed_everything"]
+
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/__init__.py:1098]:
```suggestion
```

File: tests/callbacks/test_model_checkpoint.py [hunk 14]
```diff
@@ -117,7 +117,9 @@ def test_model_checkpoint_no_extraneous_invocations(tmpdir):
     """Test to ensure that the model callback saves the checkpoints only once in distributed mode."""
     model = EvalModelTemplate()
     num_epochs = 4
-    model_checkpoint = ModelCheckpointTestInvocations(monitor='val_loss', expected_count=num_epochs, save_top_k=-1)
+    model_checkpoint = ModelCheckpointTestInvocations(monitor='val_loss',
+        expected_count=num_epochs, save_top_k=-1
+    )
```
Comment (rohitgr7) [inline tests/callbacks/test_model_checkpoint.py:122]:
```suggestion
    model_checkpoint = ModelCheckpointTestInvocations(monitor='val_loss', expected_count=num_epochs, save_top_k=-1)
```

File: tests/loggers/test_tensorboard.py [hunk 15]
```diff
@@ -35,7 +35,7 @@ def test_tensorboard_hparams_reload(tmpdir):
 
     # verify artifacts
     assert len(os.listdir(os.path.join(folder_path, "checkpoints"))) == 1
-
+
```
Comment (rohitgr7) [inline tests/loggers/test_tensorboard.py:38]:
```suggestion
```

File: tests/loggers/test_tensorboard.py [hunk 16]
```diff
@@ -49,7 +49,7 @@ def test_tensorboard_hparams_reload(tmpdir):
                   b'\x06\n\x02b2 \x03*\r\n\x0b\x12\thp_metric'
 
     hparams_data = data_pt_1_6 if LooseVersion(torch.__version__) >= LooseVersion("1.6.0") else data_pt_1_5
-
+
```
Comment (rohitgr7) [inline tests/loggers/test_tensorboard.py:52]:
```suggestion
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 17]
```diff
@@ -496,8 +496,8 @@ def _save_top_k_checkpoints(self, metrics, trainer, pl_module, epoch, filepath):
         if current is None:
             m = f"Can save best model only with {self.monitor} available, skipping."
             if self.monitor == 'checkpoint_on':
-                m = 'No checkpoint_on found. Hint: Did you set it in EvalResult(checkpoint_on=tensor) or ' \
-                    'TrainResult(checkpoint_on=tensor)?'
+                m = f'No checkpoint_on found. Hint: Did you set it in EvalResult(checkpoint_on=tensor) or ' \
+                    f'TrainResult(checkpoint_on=tensor)?'
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:500]:
```suggestion
                m = 'No checkpoint_on found. Hint: Did you set it in EvalResult(checkpoint_on=tensor) or ' \
                    'TrainResult(checkpoint_on=tensor)?'
```

**Comment by rohitgr7** [comment]
@Borda fixed the bug.
